### PR TITLE
Add HDF5 Writer Support

### PR DESCRIPTION
**Pull request by jamesm2w**
- Add a generic IO interface for fetching a writer
- Support ASCII (CSV) and HDF5 (GADGET) formats https://wwwmpa.mpa-garching.mpg.de/gadget4/06_snapshotformat/
- These can be read by `splash`
- Currently still very much working off the assumptions of the serial implementation (2 dim vectors, doubles, single type of particle, etc)
- Added new variables to Makefile to try and detect any HDF5 support. Could probably be improved in future.
- Also set MPI compiler wrappers to use clang(++)

Current usage is 
```cpp
#include "mock_io.hpp"
auto writer = wash::get_file_writer("hdf5");
writer->begin_iteration( iteration_number , "./file_path" );
```
The iteration number and extension will be added automatically to the generated files.

### CODE REVIEW

File: wash_mockapi.cpp [hunk 1]
```diff
@@ -167,4 +167,24 @@ namespace wash {
         }
     }
 
+    std::vector<std::string>* sim_get_forces_scalar() {
```
Comment (Scott943) [inline wash_mockapi.cpp:170]: Better to return a const reference here e.g. `const &std::vector<std::string>`

Reference means that the calling function cannot change the address (unlike a ptr)

If you make it const it enforces that the calling function cannot change the value of these forces.

File: wash_mockapi.cpp [hunk 2]
```diff
@@ -167,4 +167,24 @@ namespace wash {
...
+        return &forces_scalar;
+    }
+
+    std::vector<std::string>* sim_get_forces_vector() {
```
Comment (Scott943) [inline wash_mockapi.cpp:174]: Same as above

File: wash_mockapi.cpp [hunk 3]
```diff
@@ -167,4 +167,24 @@ namespace wash {
...
+        return &forces_vector;
+    }
+
+    std::vector<Particle>* sim_get_particles() {
```
Comment (Scott943) [inline wash_mockapi.cpp:178]: Same as above

File: wash_mockapi.cpp (hunk 1 above)
Comment (jamesm2w) [reply wash_mockapi.cpp]: Changes made.

**Pull request merged**